hibernate-core/src/main/java/org/hibernate/internal/util/collections/InstanceIdentityStore.java

```
@@ -0,0 +1,101 @@
+		final Page<Object> page = getPage( keyIndex );
+		if ( page != null ) {
+			final int offset = toPageOffset( keyIndex );
+			final Object k = page.get( offset );
+			if ( k == key ) {
+				//noinspection unchecked
+				return (V) page.get( offset + 1 );
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Associates the specified value with the specified key in this store (optional operation). If the store
+	 * previously contained a mapping for the key, the old value is replaced by the specified value.
+	 *
+	 * @param key key with which the specified value is to be associated
+	 * @param value value to be associated with the specified key
+	 */
+	public <K extends InstanceIdentity> void put(K key, V value) {
+		if ( key == null ) {
+			throw new NullPointerException( "This store does not support null keys" );
+		}
+
+		final int instanceId = key.$$_hibernate_getInstanceId();
+		final int keyIndex = toKeyIndex( instanceId );
+		final Page<Object> page = getOrCreateEntryPage( keyIndex );
+		final int pageOffset = toPageOffset( keyIndex );
+		page.set( pageOffset, key );
+		page.set( pageOffset + 1, value );
+	}
+
+	/**
+	 * Removes the mapping for an instance id from this store if it is present (optional operation).
+	 *
+	 * @param instanceId the instance id whose associated value is to be returned
+	 * @param key key instance to double-check instance equality
+	 * @implNote This method accesses the backing array with the provided instance id, but performs an instance
+	 * equality check ({@code ==}) with the provided key to ensure it corresponds to the mapped one
+	 */
+	public void remove(int instanceId, Object key) {
+		final int keyIndex = toKeyIndex( instanceId );
+		final Page<Object> page = getPage( keyIndex );
+		if ( page != null ) {
+			final int pageOffset = toPageOffset( keyIndex );
+			Object k = page.set( pageOffset, null );
+			// Check that the provided instance really matches with the key contained in the store
+			if ( k == key ) {
```

Maybe throw an exception when the keys don't match?